src/lib.rs

```diff
@@ -1147,8 +1150,128 @@ pub enum Expression {
     ArrayLength(Handle<Expression>),
 }
 
-/// A code block is just a vector of statements.
-pub type Block = Vec<Statement>;
+/// A code block is a vector of statements, with maybe a vector of spans.
+#[derive(Debug, Clone, Default)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize))]
+#[cfg_attr(
+    any(feature = "serialize", feature = "deserialize"),
+    serde(transparent)
+)]
+pub struct Block {
+    body: Vec<Statement>,
+    #[cfg(feature = "span")]
+    #[cfg_attr(any(feature = "serialize", feature = "deserialize"), serde(skip))]
+    span_info: Vec<Span>,
+}
+
+impl Block {
+    pub fn new() -> Self {
+        Self {
+            body: Vec::new(),
+            #[cfg(feature = "span")]
+            span_info: Vec::new(),
+        }
+    }
+
+    pub fn from_vec(body: Vec<Statement>) -> Self {
+        #[cfg(feature = "span")]
+        let span_info = std::iter::repeat(Span::Unknown).take(body.len()).collect();
+        Self {
+            body,
+            #[cfg(feature = "span")]
+            span_info,
+        }
+    }
+
+    pub fn body(&self) -> &Vec<Statement> {
+        &self.body
+    }
+
+    #[allow(unused_variables)]
+    pub fn push(&mut self, end: Statement, span: Span) {
+        self.body.push(end);
+        #[cfg(feature = "span")]
+        self.span_info.push(span);
```

Comment:
one of the risks here, I suppose, is that the number of elements in `body` diverges from the number of elements in `span_info`. Even though the fields are private, they are visible to all the submodules.
I can think of 2 approaches:
  - the lazy one - just do the check in validation
  - the strict one - move the block definition and implementation into a `block` submodule, and re-export it. This would mean the fields are really concealed, and one can at least look at the logic to ensure that both vectors are modified at the same time.

Reply:
Moved to a separate module, there's a decent amount of code now anyway.